sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2729,6 +2755,29 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * Collect a Dataset as an ArrowRecordBatch, and serve the ArrowRecordBatch to PySpark.
+   */
+  private[sql] def collectAsArrowToPython(): Int = {
+    val recordBatch = collectAsArrow()
+    val arrowSchema = Arrow.schemaToArrowSchema(this.schema)
+    val out = new ByteArrayOutputStream()
+    try {
+      val writer = new ArrowWriter(Channels.newChannel(out), arrowSchema)
+      writer.writeRecordBatch(recordBatch)
+      writer.close()
+    } catch {
+      case e: Exception =>
+        throw e
+    } finally {
+      recordBatch.close()
+    }
+
+    withNewExecutionId {
+      PythonRDD.serveIterator(Iterator(out.toByteArray), "serve-Arrow")
+    }
```

Comment:
The mechanics of taking an `ArrowPayload` (as described above) and serving it to Python could be pushed down into either `Arrow.scala` (whatever that ends up being called) or in `PythonRDD`. e.g.

```scala
val payload = collectAsArrow()
payload.serveToPython()
```